### PR TITLE
[RISCV] Generate more W instructons

### PR DESCRIPTION
**Pull request by wangpc-pp**
We rename `TuneNoStripWSuffix` to `TunePreferWInst`.

If all the users of an instruction just use the low 32 bits, we can
convert it to its W variant.

A quick test on Coremark (`-O3 -march=rv64gc`):

|        | W instructions | code size(.text) |
|--------|----------------|------------------|
| before | 302            | 12257            |
| after  | 343            | 12265            |
|        | +13.58%        | +0.065%          |

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVOptWInstrs.cpp [hunk 1]
```diff
@@ -698,6 +700,42 @@ bool RISCVOptWInstrs::stripWSuffixes(MachineFunction &MF,
...
+      case RISCV::MUL:
+        Opc = RISCV::MULW;
+        break;
+      case RISCV::SLLI:
+        Opc = RISCV::SLLIW;
+        break;
+      }
+
+      if (hasAllWUsers(MI, ST, MRI)) {
+        MI.setDesc(TII.get(Opc));
```
Comment (michaelmaitland) [inline llvm/lib/Target/RISCV/RISCVOptWInstrs.cpp:730]: Is it possible that `Opc` is uninitialized but used here?

Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVOptWInstrs.cpp:730]: I don't think so, there is a `continue` for the default case.

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCVOptWInstrs.cpp:730]: Yes, my bad.

Comment (michaelmaitland) [inline llvm/lib/Target/RISCV/RISCVOptWInstrs.cpp:730]: Add `NumTransformedToWInstrs++`?

File: llvm/lib/Target/RISCV/RISCVOptWInstrs.cpp [hunk 2]
```diff
@@ -46,6 +46,9 @@ static cl::opt<bool> DisableSExtWRemoval("riscv-disable-sextw-removal",
 static cl::opt<bool> DisableStripWSuffix("riscv-disable-strip-w-suffix",
                                          cl::desc("Disable strip W suffix"),
                                          cl::init(false), cl::Hidden);
+static cl::opt<bool> EnableAppendWSuffix("riscv-enable-append-w-suffix",
```
Comment (dtcxzyw) [inline llvm/lib/Target/RISCV/RISCVOptWInstrs.cpp:49]: Should we update the file header?

File: llvm/lib/Target/RISCV/RISCVOptWInstrs.cpp [hunk 3]
```diff
@@ -698,6 +709,58 @@ bool RISCVOptWInstrs::stripWSuffixes(MachineFunction &MF,
...
+        WOpc = RISCV::LW;
+        break;
+      }
+
+      if (hasAllWUsers(MI, ST, MRI)) {
+        LLVM_DEBUG(dbgs() << "Replacing " << MI);
+        MI.setDesc(TII.get(WOpc));
+        MI.clearFlag(MachineInstr::MIFlag::NoSWrap);
+        MI.clearFlag(MachineInstr::MIFlag::NoUWrap);
+        MI.clearFlag(MachineInstr::MIFlag::IsExact);
```
Comment (dtcxzyw) [inline llvm/lib/Target/RISCV/RISCVOptWInstrs.cpp:753]: Can you add some tests for flags dropping?

File: llvm/lib/Target/RISCV/RISCVOptWInstrs.cpp [hunk 4]
```diff
@@ -711,7 +774,12 @@ bool RISCVOptWInstrs::runOnMachineFunction(MachineFunction &MF) {
 
   bool MadeChange = false;
   MadeChange |= removeSExtWInstrs(MF, TII, ST, MRI);
-  MadeChange |= stripWSuffixes(MF, TII, ST, MRI);
+
+  if (!(DisableStripWSuffix || ST.preferWInst()))
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVOptWInstrs.cpp:778]: These flags should not override the target preferences. DisableStripWSuffix was intended to disable the optimization if we found a bug.

File: llvm/lib/Target/RISCV/RISCVOptWInstrs.cpp [hunk 5]
```diff
@@ -711,7 +774,12 @@ bool RISCVOptWInstrs::runOnMachineFunction(MachineFunction &MF) {
...
+    MadeChange |= stripWSuffixes(MF, TII, ST, MRI);
+
+  if (EnableAppendWSuffix || ST.preferWInst())
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVOptWInstrs.cpp:781]: `EnableAppendWSuffix` should not override the target preference. The command line option should only be for debug if the pass happens to introduce a bug. You can use -mattr=+prefer-w-inst for testing.

File: llvm/lib/Target/RISCV/RISCVOptWInstrs.cpp (hunk 3 above)
Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVOptWInstrs.cpp:753]: It's the same as `RISCVOptWInstrs::removeSExtWInstrs`:
https://github.com/llvm/llvm-project/blob/a8c59750d911eb30d5664696db19af445dd770f8/llvm/lib/Target/RISCV/RISCVOptWInstrs.cpp#L648-L657
I don't know how to test it actually.

**Comment by wangpc-pp** [comment]
Ping.

**Review by topperc — APPROVED** [review]
LGTM

**Review by dtcxzyw — APPROVED** [review]
LGTM.

**Pull request merged**